### PR TITLE
Extensibility improvements

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# Extensibility improvements

This module has some vintage SS 3 code which has hardcoded statically accessed classes. This makes extensibility difficult. Following changes were made

* all public static API replaced with Injectable version
* legacy public static API still available via deprecated methods
* most `ErrorPage` functionality moved to an extension so it's easy to create a different page type with error page functionality
* documentation updated on how to inject over existing error page with the most notable use case - an error page made out of blocks
* minor code cleanups (moving hard coded values to constants, fixing outdates comments...)

## Backwards compatibility

* no behaviour change
* no configuration API change
* no extension points change
* there are lot of API changes but a legacy API is still supported

## Related issues

https://github.com/silverstripe/silverstripe-errorpage/issues/44

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -75,7 +75,105 @@ ErrorPage:
...
+     * @var array
+     */
+    private static $extensions = [
+        ErrorPageExtension::class,
+    ];
+
+    /**
+     * @return FieldList
+     */
+    public function getCMSFields(): FieldList
```
Comment (unclecheese) [inline README.md]: Won't this throw a strict warning? The parent class doesn't declare an explicit return type AFAIK.

File: src/ErrorPageExtension.php [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+     */
+    public function defaultRecordsAllowed()
+    {
+        return true;
+    }
+
+    /**
+     * @return DropdownField
+     */
+    public function createErrorCodeField()
```
Comment (unclecheese) [inline src/ErrorPageExtension.php]: Explicit returns on these two methods, please ^

File: src/ErrorPageExtension.php [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+                'created'
+            );
+        } else {
+            DB::alteration_message(
+                sprintf('%s error page created', $code),
+                'created'
+            );
+        }
+    }
+}
```
Comment (unclecheese) [inline src/ErrorPageExtension.php]: This whole extension is a new class, so let's feel free to PHP7 it up with typing. :)

File: README.md (hunk 1 above)
Comment (mfendeksilverstripe) [reply README.md]: I tested this and it doesn't seem to be a problem with that. The issue would be only if this was an `Interface` function.

**Comment by mfendeksilverstripe** [comment]
Replaced by https://github.com/silverstripe/silverstripe-errorpage/pull/45

**Pull request closed without merging**